### PR TITLE
feat: ContractAuthorization (authz)

### LINKED ISSUE #803: Authz module integration - more granularity for WasmExecuteMsg authorizations
Currently, it's possible to grant permission for an address to perform the `WasmExecuteMsg` against _any_ contract. This not a very useful granularity.

It would be nice to grant an account permission to execute only against a specific contract. Even better, if the message could be optionally specified.

This could be useful for DAOs, that want to grant a user permission to execute a contract on behalf of the DAO. For example, granting a user (or cron bot) permission to claim rewards on behalf of the DAO.

### PR DESCRIPTION
**Pull request by giansalex**
closes #803

cli examples:
```
CONTRACT=wasm14hj2tavq8fpesdwxxcu44rty3hh90vhujrvcmstl4zr3txmfvw9s0phg4d
GRANTEE=wasm14vhcdsyf83ngsrrqc92kmw8q9xakqjm048cr0y

wasmd tx wasm grant $GRANTEE $CONTRACT --allow-msgs claim,bond --from granter
wasmd tx authz revoke $GRANTEE /cosmwasm.wasm.v1.MsgExecuteContract --from granter
```

query grants
```
grants:
- authorization:
    '@type': /cosmwasm.wasm.v1.ContractAuthorization
    contract: wasm14hj2tavq8fpesdwxxcu44rty3hh90vhujrvcmstl4zr3txmfvw9s0phg4d
    messages:
    - claim
    - bond 
    once: false
  expiration: "2024-08-28T02:03:09Z"
  grantee: wasm14vhcdsyf83ngsrrqc92kmw8q9xakqjm048cr0y
  granter: wasm1tr944mwl56j7yhqkgq5xmah8w7udjwpmj33a4y
pagination:
  next_key: null
  total: "0"
  ```

### CODE REVIEW

**Review by alpe — COMMENTED** [review]
Thanks a lot for the PR and driving this forward. Very nice! I had a quick look and added some comments.

I assume the authz module ensures that the sender has permission to set the ContractAutorization. But how does it work when not called from the contract? The CLI does not have keys for the contract address. I need to 👁️ at the module but if you can point me to the code line , that would help

File: proto/cosmwasm/wasm/v1/authz.proto [hunk 1]
```diff
@@ -0,0 +1,21 @@
...
+  option (cosmos_proto.implements_interface) = "Authorization";
+
+  // Contract is the address of the smart contract
+  string contract = 1;
+
+  // Messages is the list of messages that can be executed
+  repeated string messages = 2;
+
+  // Once specifies if the contract can only be called once
+  bool once = 3;
```
[the remark this answers is not in the document]
Comment (giansalex) [reply proto/cosmwasm/wasm/v1/authz.proto:20]: yes, can be `times` ?

File: x/wasm/types/authz.go [hunk 2]
```diff
@@ -0,0 +1,59 @@
...
+		return authztypes.AcceptResponse{}, sdkerrors.ErrInvalidType.Wrap("type mismatch")
+	}
+
+	if a.Contract != exec.Contract {
+		return authztypes.AcceptResponse{}, sdkerrors.ErrUnauthorized.Wrapf("cannot run %s contract", exec.Contract)
+	}
+
+	gasForDeserialization := gasDeserializationCostPerByte * uint64(len(exec.Msg))
+	ctx.GasMeter().ConsumeGas(gasForDeserialization, "contract authorization")
+	if err := IsJSONObjectWithTopLevelKey(exec.Msg, a.Messages); err != nil {
```
[the remark this answers is not in the document]
Comment (giansalex) [reply x/wasm/types/authz.go:40]: I think `IsJSONObjectWithTopLevelKey` verifies the requirements.

_"All contract messages"_ is covered with default authz behavior, using generic authorization

```
wasmdtx authz grant wasm14vhcdsyf83ngsrrqc92kmw8q9xakqjm048cr0y \
  generic --msg-type /cosmwasm.wasm.v1.MsgExecuteContract --from granter
```

**Comment by giansalex** [comment]
Any user can grant ContractAutorization.

Full example:

User grants the "claim" msg
```bash
wasmd tx wasm grant $GRANTEE_ADDR $STAKE_CONTRACT --allow-msgs claim --from granter
```

Grantee executes the contract
```bash
wasmd tx wasm execute $STAKE_CONTRACT '{"claim":{}}' --from $GRANTER_ADDR --generate-only > claim.json

wasmd tx authz exec claim.json --from grantee
```

**Comment by giansalex** [comment]
I need to add a list in ContractAuthorization, because with the current implementation I can only give permission to a single contract per grantee.

eg. I could not give permissions ContractX:claim and ContractY:swap to the same grantee.

**Comment by alpe** [comment]
Thanks for the examples! This helped me to get a better picture! 

>I need to add a list in ContractAuthorization, because with the current implementation I can only give permission to a single contract per grantee.

Good finding. The store key is built from [grantee, granter, msgType](https://github.com/cosmos/cosmos-sdk/blob/v0.45.8/x/authz/keeper/keys.go#L23).
Adding a list of contract addresses would be a simple solution but this enables ` ContractX:claim,swap  and ContractY:claim,swap` while you probably want `ContractX:claim and ContractY:swap`. 
With a wildcard things get even more in the wrong direction. I think it makes sense to extract the contract, keys tuple into a new proto type and iterate over the list to find the matching contract authorization. WDYT?

**Comment by alpe** [comment]
@giansalex I have started a spike based on your work to model this. I may be able to push this later today or on Monday. The proto would looks like below. WDYT?

```proto
// ContractExecutionAuthorization defines authorization for wasm execute.
message ContractExecutionAuthorization {
  option (cosmos_proto.implements_interface) = "Authorization";

  repeated ContractExecutionGrant grants = 1 [ (gogoproto.nullable) = false ];
}

// ContractExecutionGrant a granted execute permission for a single contract
message ContractExecutionGrant {
  // Contract is the address of the smart contract
  string contract = 1;

  // Filter rules to apply
  oneof filter {
    AcceptedMessageKeysFilter accepted_message_keys = 2;
    AllowAllWildcard allow_all_wildcard = 3;
  }

  // RemainingCalls specifies the number of authorized executions remaining
  uint64 remaining_calls = 4;
}

// AllowAllWildcard is a wildcard to allow any type of contract execution
// message
message AllowAllWildcard {}

// AcceptedMessageKeysFilter accept specific contract message keys in the json
// object that can be executed
message AcceptedMessageKeysFilter {
  // Messages is the list of unique keys
  repeated string messages = 1;
}
```

**Comment by giansalex** [comment]
> I think it makes sense to extract the contract, keys tuple into a new proto type and iterate over the list to find the matching contract authorization.

Yes, I was also thinking about a list.

> @giansalex I have started a spike based on your work to model this. I may be able to push this later today or on Monday. The proto would looks like below. WDYT?

About remaining_calls, what happens when it becomes unlimited, will 0 be used?

**Comment by alpe** [comment]
> About remaining_calls, what happens when it becomes unlimited, will 0 be used?

good point! IMHO it makes sense to make it explicit or do not support unlimited. I have pushed an example to:
https://github.com/CosmWasm/wasmd/tree/966_grants_spike

**Comment by giansalex** [comment]
i am not against _unlimited_ because authz MsgGrant has an expiration field, although with a high value in `remaining_calls` may be enough.

I did not see any new changes in [966_grants_spike](https://github.com/CosmWasm/wasmd/tree/966_grants_spike) branch

**Comment by ethanfrey** [comment]
These features will be in the next v0.30 release, maybe the version in #978 based on it.

Let's work together to get some clear design we all agree on.

File: proto/cosmwasm/wasm/v1/authz.proto [hunk 3]
```diff
@@ -0,0 +1,21 @@
...
+
+// ContractAuthorization defines authorization for wasm execute.
+message ContractAuthorization {
+  option (cosmos_proto.implements_interface) = "Authorization";
+
+  // Contract is the address of the smart contract
+  string contract = 1;
+
+  // Messages is the list of messages that can be executed
+  repeated string messages = 2;
```
[the remark this answers is not in the document]
Comment (mikedotexe) [reply proto/cosmwasm/wasm/v1/authz.proto:17]: Are the messages the same as "execute methods" here? Amazing work, just catching up now

**Pull request closed without merging**